exec/java-exec/src/main/java/org/apache/drill/exec/rpc/user/UserServer.java

```diff
@@ -308,33 +312,57 @@ public BitToUserHandshake getHandshakeResponse(UserToBitHandshake inbound) throw
         connection.setHandshake(inbound);
 
         try {
+          // TODO(SUDHEESH): MUST FIX THIS VERSION CHECK FIRST BEFORE THE CHECK BELOW
           if (inbound.getRpcVersion() != UserRpcConfig.RPC_VERSION) {
             final String errMsg = String.format("Invalid rpc version. Expected %d, actual %d.",
                 UserRpcConfig.RPC_VERSION, inbound.getRpcVersion());
 
             return handleFailure(respBuilder, HandshakeStatus.RPC_VERSION_MISMATCH, errMsg, null);
           }
 
-          if (authenticator != null) {
-            try {
-              String password = "";
-              final UserProperties props = inbound.getProperties();
-              for (int i = 0; i < props.getPropertiesCount(); i++) {
-                Property prop = props.getProperties(i);
-                if (UserSession.PASSWORD.equalsIgnoreCase(prop.getKey())) {
-                  password = prop.getValue();
-                  break;
+          connection.setHandshake(inbound);
+
+          if (authFactory != null) {
+            if (inbound.getRpcVersion() <= 5) { // for backward compatibility <= 1.8
+              final String userName = inbound.getCredentials().getUserName();
+              if (logger.isTraceEnabled()) {
+                logger.trace("User {} on connection {} is using an older client (Drill version <= 1.8).",
+                    userName, connection.getRemoteAddress());
+              }
+              try {
+                String password = "";
+                final UserProperties props = inbound.getProperties();
+                for (int i = 0; i < props.getPropertiesCount(); i++) {
+                  Property prop = props.getProperties(i);
+                  if (UserSession.PASSWORD.equalsIgnoreCase(prop.getKey())) {
+                    password = prop.getValue();
+                    break;
+                  }
+                }
+                final PlainMechanism plainMechanism = authFactory.getPlainMechanism();
+                if (plainMechanism == null) {
+                  throw new UserAuthenticationException("The server no longer supports username/password" +
+                      " based authentication. Please talk to your system administrator.");
                 }
+                plainMechanism.getAuthenticator().authenticate(userName, password);
+                connection.changeHandlerTo(handler);
+                connection.finalizeSession(userName);
+                respBuilder.setStatus(HandshakeStatus.SUCCESS);
```

Comment:
Would it be useful to add log the successful mechanism (where we have `respBuilder.setStatus(HandshakeStatus.SUCCESS);`)?